### ISSUE TITLE
How is the information passed to a keras loss function, with Theano backend

### ISSUE THREAD

**Issue by michallhal** [fchollet/keras#4327]
I am fairly new to keras and DL and I am trying to build a loss function but I have questions about how the data from my network is passed through y_pred and y_true of the loss function.

As an example, my network has 3 different outputs here is one:

SEC5 = merge( [SEC1_up, SEC2_up, SEC3_up, SEC4_up], mode='concat', concat_axis=1 )
SEC5 = Convolution2D( 2,1,1, subsample=(1, 1), border_mode='same', activation="sigmoid" )( SEC5 )

SEC5 is now a 2 channel tensor that is predicting edges in one channel and non-edges in the other.

My model is created with the following line:
model = Model( input=inputs, output=[Final, ILLP2, SEC5] )

Where I perform binary cross entropy on Final, Squared loss on ILLP2, and then a custom loss for the SEC layer. When building the custom loss I have come across something that I don't understand. How are multiple channel layers (like SEC5) passed to the loss function? This is particularly important in my edge loss as I need to calculate the number of edges in the edge layer, and the number of non edges in the non edge layer.

What I don't understand is the actual variable in the loss function (y_true and y_pred) when I do this:

print 'y_true data'
print y_true.ndim
print y_true.type
print 'y_pred data'
print y_pred.ndim
print y_pred.type

I get the following values (which make sense to me):

y_true data
2
TensorType(float32, matrix)
y_pred data
2
TensorType(float32, matrix)

And this is where i get really confused by everything. As I understand it, tensortypes of matrix can only be 2 dimensional, but I essentially have 3 dimensions? How does it deal with this information?

I feel like I should understand this before I go making elaborate loss functions of my own, any information you could provide me with would be greatly appreciated.

Cheers,

Michael

**Comment by pechersky** [comment]
When you fit or train your model, what shape is your "true output data"? Do you have a separate set for each of the `[Final, ILLP2, SEC5]`?

**Comment by michallhal** [comment]
Hi pechersky, 
My true data is as follows, Final [NumbeData, 18, 104, 152], ILLP2 [NumberData, 18], and SEC is [NumberData, 2, 104, 152]

**Comment by alyato** [comment]
@michallhal ,Does you solve it? thanks.
